spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java

```diff
@@ -178,12 +179,16 @@ private SparkTableUtil() {
 
       Option<scala.collection.immutable.Map<String, String>> scalaPartitionFilter;
       if (partitionFilter != null && !partitionFilter.isEmpty()) {
-        scalaPartitionFilter = Option.apply(JavaConverters.mapAsScalaMapConverter(partitionFilter).asScala()
-            .toMap(Predef.conforms()));
+        Builder<Tuple2<String, String>, scala.collection.immutable.Map<String, String>> builder =
+            Map$.MODULE$.<String, String>newBuilder();
+        for (Map.Entry<String, String> entry : partitionFilter.entrySet()) {
+          builder.$plus$eq(Tuple2.apply(entry.getKey(), entry.getValue()));
+        }
```

Comment:
Using the Scala API in Java is really ugly. I think that's why we prefer to use the Java API and convert unless it is performance critical. Is this change needed for Scala 2.13 or can we continue to convert?

Reply:
I fully agree with you that "using the Scala API in Java is really ugly”, but in this case, I fear this is the better choice. 

I thought it would be really simple, all we want to do is convert a java Map into an *immutable* Scala map. 
So, we start off by converting a java Map into a *mutable* Scala map:  
> scala.collection.mutable.Map<String, String> myMutableMap = JavaConverters.mapAsScalaMapConverter(myJavaMap).asScala();`

And now we can convert a *mutable* Scala map into an *immutable* Scala map using the handy-dandy `.toMap` method. 
But wait, the `.toMap` method takes an implicit argument `ev`: 
> def toMap[K, V](implicit ev: A <:< (K, V)): immutable.Map[K, V]

This is basically a compiler trick to make sure we’re calling `.toMap` on a collection of tuples (otherwise the method doesn’t make sense). 
Normally, in Scala, this evidence is implicitly supplied. 
Since we’re calling this method from Java, we don’t have the normal implicit mechanisms available so we have to construct and supply this evidence by hand. 
In the past we were able to supply this evidence by using `scala.Predef.conforms()` but this API was removed entirely in 2.13 (it was already deprecated in 2.12). 
AFAICT, in scala 2.13, this implicit `ev` value of type `<:<` is provided by `scala.Predef.refl()` in in Scala 2.13. 
However, this method is not available in scala 2.12. 
Right now, the only other alternative I can think of is having two separate files for this code for 2.12 and 2.13, but that seems overkill for something so small, hence I would argue this is the better choice. 

Hope that makes sense. Open to thoughts.

Reply:
I've added one more change [here](https://github.com/apache/iceberg/pull/4009/commits/440bc10b89dfe85893c7e8d2b150b6a491cf73bc) so at least it looks a little less gnarly :)

Reply:
Okay, so the problem is that we can't convert directly to an Immutable Scala map, and the Scala API's use of implicits makes converting mutable to immutable really difficult. That seems reasonable. I'm not sure what we were doing before... maybe we didn't need an Immutable map?